### PR TITLE
Refactor CCS handling code from EsqlSession and IndexResolver into dedicated util class

### PR DESCRIPTION
**Pull request by quux00**
This a pure refactoring ticket. No new functionality was added. 
A few new tests were added since they are exposed in a place for easier testing.
Based on https://github.com/elastic/elasticsearch/pull/115266#pullrequestreview-2403485393

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java [hunk 1]
```diff
@@ -150,103 +143,15 @@ public void execute(
...
-                if (e instanceof ConnectTransportException) {
-                    // when field-caps has no field info (since no clusters could be connected to or had matching indices)
-                    // it just throws the first exception in its list, so this odd special handling is here is to avoid
-                    // having one specific remote alias name in all failure lists in the metadata response
-                    exceptionForResponse = new RemoteTransportException(
-                        "connect_transport_exception - unable to connect to remote cluster",
-                        null
-                    );
+            ActionListener.wrap(plan -> executeOptimizedPlan(request, executionInfo, runPhase, optimizedPlan(plan), listener), e -> {
+                if (EsqlSessionCCSUtils.returnSuccessWithEmptyResult(executionInfo, e)) {
```
Comment (smalyshev) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:147]: I wonder if we could go further and put this logic into `EsqlSessionCCSUtils` too. So we'd have just a method name here. After all, creating an empty result is also a generic function that is not very specific to EsqlSession?

**Review by smalyshev — APPROVED** [review]

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSessionCCSUtils.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+    private EsqlSessionCCSUtils() {}
+
+    static String createIndexExpressionFromAvailableClusters(EsqlExecutionInfo executionInfo) {
+        StringBuilder sb = new StringBuilder();
+        for (String clusterAlias : executionInfo.clusterAliases()) {
+            EsqlExecutionInfo.Cluster cluster = executionInfo.getCluster(clusterAlias);
+            if (cluster.getStatus() != EsqlExecutionInfo.Cluster.Status.SKIPPED) {
+                if (cluster.getClusterAlias().equals(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY)) {
+                    sb.append(executionInfo.getCluster(clusterAlias).getIndexExpression()).append(',');
```
Comment (pawankartik-elastic) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSessionCCSUtils.java:38]: `executionInfo.getCluster(clusterAlias).getIndexExpression()` is called in both the `if` branches. I wonder if it'd be better to pull it out?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSessionCCSUtils.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                } else {
+                    String indexExpression = executionInfo.getCluster(clusterAlias).getIndexExpression();
+                    for (String index : indexExpression.split(",")) {
+                        sb.append(clusterAlias).append(':').append(index).append(',');
+                    }
+                }
+            }
+        }
+
+        if (sb.length() > 0) {
```
Comment (pawankartik-elastic) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSessionCCSUtils.java:48]: Suggestion: use `isEmpty()`?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSessionCCSUtils.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        }
+
+        if (e instanceof NoClustersToSearchException || ExceptionsHelper.isRemoteUnavailableException(e)) {
+            for (String clusterAlias : executionInfo.clusterAliases()) {
+                if (executionInfo.isSkipUnavailable(clusterAlias) == false
+                    && clusterAlias.equals(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY) == false) {
+                    return false;
+                }
+            }
+            return true;
```
Comment (pawankartik-elastic) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSessionCCSUtils.java:170]: What's your opinion on writing this as:
```java
return executionInfo.clusterAliases()
    .stream()
    .noneMatch(clusterAlias -> executionInfo.isSkipUnavailable(clusterAlias) == false && clusterAlias.equals(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY) == false
);
```

**Comment by quux00** [comment]
This refactoring overlaps a lot of the refactoring Costin did in his PR around lookups: https://github.com/elastic/elasticsearch/pull/115813#pullrequestreview-2406675530

so probably makes sense to close this PR and wait for his to merge.

**Comment by quux00** [comment]
Closing since Costin has done much of this refactoring in https://github.com/elastic/elasticsearch/pull/115813.
I will address a few other minor refactorings not included in his PR in my next ticket around CCS ESQL skip_unavailable.

**Pull request closed without merging**